WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs

```diff
@@ -339,33 +339,33 @@ private void BlockNotifier_OnBlock(object? sender, Block e)
 			currentIndex = Index;
 		}
 
-		foreach (var filter in currentIndex)
+		if (currentIndex.Count == 0)
 		{
-			if (found)
-			{
-				filters.Add(filter);
-				if (filters.Count >= count)
-				{
-					break;
-				}
-			}
-			else
-			{
-				if (filter.Header.BlockHash == bestKnownBlockHash)
-				{
-					found = true;
-				}
-			}
+			return (Height.Unknown, []);
 		}
 
-		if (currentIndex.Count == 0)
+		// Search for bestKnownBlockHash from last to first
+		var i = currentIndex.Count - 1;
+		for (; i >= 0; i--)
 		{
-			return (Height.Unknown, Enumerable.Empty<FilterModel>());
+			if (!currentIndex[i].Header.BlockHash.Equals(bestKnownBlockHash))
+			{
+				continue;
+			}
+
+			found = true;
+			break;
 		}
-		else
+
+		if (found && i < currentIndex.Count - 1)
 		{
-			return ((int)currentIndex[^1].Header.Height, filters);
+			// Populate filters starting from the found index + 1
+			var startIndex = i + 1;
```

Comment:
It is not so intuitive what's happening here but it's because we don't want the filter of the block we already know, we want the next one